drivers/java/driver.go

```diff
@@ -0,0 +1,564 @@
+
+	// logger will log to the plugin output which is usually an 'executor.out'
+	// file located in the root of the TaskDir
+	logger hclog.Logger
+}
+
+func NewDriver(logger hclog.Logger) drivers.DriverPlugin {
+	ctx, cancel := context.WithCancel(context.Background())
+	logger = logger.Named(pluginName)
+	return &Driver{
+		eventer:        eventer.NewEventer(ctx, logger),
+		tasks:          newTaskStore(),
+		ctx:            ctx,
+		signalShutdown: cancel,
+		logger:         logger,
+	}
+}
+
+func (d *Driver) PluginInfo() (*base.PluginInfoResponse, error) {
+	return pluginInfo, nil
+}
+
+func (d *Driver) ConfigSchema() (*hclspec.Spec, error) {
+	return configSpec, nil
+}
+
+func (d *Driver) SetConfig(_ []byte, cfg *base.ClientAgentConfig) error {
+	if cfg != nil {
+		d.nomadConfig = cfg.Driver
+	}
+	return nil
+}
+
+func (d *Driver) TaskConfigSchema() (*hclspec.Spec, error) {
+	return taskConfigSpec, nil
+}
+
+func (d *Driver) Capabilities() (*drivers.Capabilities, error) {
+	return capabilities, nil
+}
+
+func (r *Driver) Fingerprint(ctx context.Context) (<-chan *drivers.Fingerprint, error) {
+	ch := make(chan *drivers.Fingerprint)
+	go r.handleFingerprint(ctx, ch)
+	return ch, nil
+}
+
+func (d *Driver) handleFingerprint(ctx context.Context, ch chan *drivers.Fingerprint) {
+	ticker := time.NewTimer(0)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-d.ctx.Done():
+			return
+		case <-ticker.C:
+			ticker.Reset(fingerprintPeriod)
+			ch <- d.buildFingerprint()
+		}
+	}
+}
+
+func (d *Driver) buildFingerprint() *drivers.Fingerprint {
+	fp := &drivers.Fingerprint{
+		Attributes:        map[string]string{},
+		Health:            drivers.HealthStateHealthy,
+		HealthDescription: "healthy",
+	}
+
+	if runtime.GOOS == "linux" {
+		// Only enable if w are root and cgroups are mounted when running on linux system
+		if syscall.Geteuid() != 0 {
+			fp.Health = drivers.HealthStateUnhealthy
+			fp.HealthDescription = "java driver must run as root"
+			return fp
+		}
+
+		mount, err := fingerprint.FindCgroupMountpointDir()
+		if err != nil {
+			fp.Health = drivers.HealthStateUnhealthy
+			fp.HealthDescription = "failed to discover cgroup mount point"
+			d.logger.Warn(fp.HealthDescription, "error", err)
+			return fp
+		}
+
+		if mount == "" {
+			fp.Health = drivers.HealthStateUnhealthy
+			fp.HealthDescription = "cgroups are unavailable"
+			return fp
+		}
+	}
+
+	version, runtime, vm, err := javaVersionInfo()
+	if err != nil {
+		// return no error, as it isn't an error to not find java, it just means we
+		// can't use it.
+		fp.Health = drivers.HealthStateUndetected
+		fp.HealthDescription = ""
+		return fp
+	}
+
+	fp.Attributes[driverAttr] = "1"
+	fp.Attributes[driverVersionAttr] = version
+	fp.Attributes["driver.java.runtime"] = runtime
+	fp.Attributes["driver.java.vm"] = vm
+
+	return fp
+}
+
+func (d *Driver) RecoverTask(handle *drivers.TaskHandle) error {
+	if handle == nil {
+		return fmt.Errorf("error: handle cannot be nil")
+	}
+
```

Comment:
Check other drivers on master for the `if _, ok := Get(handle); ok { return nil }` noop fastpath. Otherwise you may avoid leaking a connection to an existing task executor (not a big deal, but still a small resource leak).

Reply:
Argh, my bad. That change is in #4510, not master: https://github.com/hashicorp/nomad/pull/4828/files#diff-cbc0ef8052eab2f7e7298b703bd429dbR281

Since that isn't merged yet we can wait on implementing it here.

Sorry for the noise.

Reply:
Updated the restore in https://github.com/hashicorp/nomad/pull/4825/commits/62d24ce9effb1b27cf5c8cae8293e8fd0a931519